File: app/src/main/java/tech/pauly/findapet/dependencyinjection/ApplicationModule.java
package tech.pauly.findapet.dependencyinjection;

import android.app.Application;
import android.content.Context;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import io.reactivex.Scheduler;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

@Module
class ApplicationModule {
    private Application application;

    public ApplicationModule(Application application) {
        this.application = application;
    }

    @Provides
    @ForApplication
    Context provideContext() {
        return application;
    }

    @Provides
    @IoScheduler
    Scheduler provideIoScheduler() {
        return Schedulers.io();
    }

    @Provides
    @Singleton
    @MainThreadScheduler
    Scheduler provideMainScheduler() {
        return AndroidSchedulers.mainThread();
    }
}

File: app/src/main/java/tech/pauly/findapet/dependencyinjection/ApplicationModule.kt
package tech.pauly.findapet.dependencyinjection

import android.app.Application
import android.content.Context

import javax.inject.Singleton

import dagger.Module
import dagger.Provides
import io.reactivex.Scheduler
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers

@Module
internal class ApplicationModule(private val application: Application) {

    @Provides
    @ForApplication
    fun provideContext(): Context {
        return application
    }

    @Provides
    @IoScheduler
    fun provideIoScheduler(): Scheduler {
        return Schedulers.io()
    }

    @Provides
    @Singleton
    @MainThreadScheduler
    fun provideMainScheduler(): Scheduler {
        return AndroidSchedulers.mainThread()
    }
}
